pkg/controller/syncsetinstance/syncsetinstance_controller.go

```diff
@@ -0,0 +1,714 @@
+
+func (r *ReconcileSyncSetInstance) handleClusterDeployment(a handler.MapObject) []reconcile.Request {
+	cd, ok := a.Object.(*hivev1.ClusterDeployment)
+	if !ok {
+		return []reconcile.Request{}
+	}
+
+	syncSetInstanceList := &hivev1.SyncSetInstanceList{}
+	err := r.List(context.TODO(), &client.ListOptions{Namespace: cd.Namespace}, syncSetInstanceList)
+	if err != nil {
+		r.logger.WithError(err).Error("cannot list syncSetInstances for cluster deployment")
+		return []reconcile.Request{}
+	}
+
+	retval := []reconcile.Request{}
+	for _, syncSetInstance := range syncSetInstanceList.Items {
+		if metav1.IsControlledBy(&syncSetInstance, cd) {
+			retval = append(retval, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      cd.Name,
+					Namespace: cd.Namespace,
+				},
+			})
+		}
+	}
+	return retval
+}
+
+var _ reconcile.Reconciler = &ReconcileSyncSetInstance{}
+
+// ReconcileSyncSetInstance reconciles a ClusterDeployment and the SyncSets associated with it
+type ReconcileSyncSetInstance struct {
+	client.Client
+	scheme *runtime.Scheme
+
+	logger               log.FieldLogger
+	applierBuilder       func([]byte, log.FieldLogger) Applier
+	hash                 func([]byte) string
+	dynamicClientBuilder func(string) (dynamic.Interface, error)
+}
+
+// Reconcile applies SyncSet or SelectorSyncSets associated with SyncSetInstances to the owning cluster.
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=syncsetinstances,verbs=get;create;update;delete;patch;list;watch
+func (r *ReconcileSyncSetInstance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the ClusterDeployment instance
+	ssi := &hivev1.SyncSetInstance{}
+
+	err := r.Get(context.TODO(), request.NamespacedName, ssi)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request
+		r.logger.WithError(err).Error("error looking up syncsetinstance")
+		return reconcile.Result{}, err
+	}
+
+	ssiLog := r.logger.WithField("syncsetinstance", request.NamespacedName)
+
+	if ssi.DeletionTimestamp != nil {
+		if !controllerutils.HasFinalizer(ssi, hivev1.FinalizerSyncSetInstance) {
+			return reconcile.Result{}, nil
+		}
+		return r.syncDeletedSyncSetInstance(ssi, ssiLog)
+	}
+
+	if !controllerutils.HasFinalizer(ssi, hivev1.FinalizerSyncSetInstance) {
+		ssiLog.Debugf("adding  finalizer")
+		if err := r.addSyncSetInstanceFinalizer(ssi); err != nil {
+			ssiLog.WithError(err).Error("error adding finalizer")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	cd := &hivev1.ClusterDeployment{}
+	cdName := types.NamespacedName{Namespace: ssi.Namespace, Name: ssi.Spec.ClusterDeployment.Name}
+	ssiLog = ssiLog.WithField("clusterdeployment", cdName)
+	err = r.Get(context.TODO(), cdName, cd)
+	if err != nil {
+		ssiLog.WithError(err).Error("error looking up clusterdeployment")
+		return reconcile.Result{}, nil
+	}
```

Comment:
Add the unreachable check and skip from: https://github.com/openshift/hive/pull/438